crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs

```diff
@@ -0,0 +1,529 @@
+        //     items: vec!['c'],
+        // }
+        // Update::ReattachItemsDone
+        // ```
+        //
+        // To ensure an optimised behaviour of this algorithm:
+        //
+        // * `Update::DetachLastItems` must not emit `VectorDiff::Remove`,
+        //
+        // * `Update::PushItems` must not emit `VectorDiff::Insert`s or
+        //   `VectorDiff::Append`s if it happens after `ReattachItems` and before
+        //   `ReattachItemsDone`. However, `Self::chunks` must always be updated.
+        //
+        // From the `VectorDiff` “point of view”, this optimisation aims at avoiding
+        // removing items to push them again later.
+        let mut mute_push_items = false;
+
+        for update in updates {
+            match update {
+                Update::NewItemsChunk { previous, new, next }
+                | Update::NewGapChunk { previous, new, next, .. } => {
+                    match (previous, next) {
+                        // New chunk at the end.
+                        (Some(previous), None) => {
+                            debug_assert!(
+                                matches!(self.chunks.back(), Some((p, _)) if p == previous),
+                                "Inserting new chunk at the end: The previous chunk is invalid"
+                            );
+
+                            self.chunks.push_back((*new, 0));
+                        }
+
+                        // New chunk at the beginning.
+                        (None, Some(next)) => {
+                            debug_assert!(
+                                matches!(self.chunks.front(), Some((n, _)) if n == next),
+                                "Inserting new chunk at the end: The previous chunk is invalid"
+                            );
+
+                            self.chunks.push_front((*new, 0));
+                        }
+
+                        // New chunk is inserted between 2 chunks.
+                        (Some(previous), Some(next)) => {
+                            let next_chunk_index = self
+                                .chunks
+                                .iter()
+                                .position(|(chunk_identifier, _)| chunk_identifier == next)
+                                // SAFETY: Assuming `LinkedChunk` and `Updates` are not buggy, and
+                                // assuming `Self::chunks` is correctly initialized, it is not
+                                // possible to insert a chunk between two chunks where one does not
+                                // exist. If this predicate fails, it means `LinkedChunk` or
+                                // `Updates` contain a bug.
+                                .expect("Inserting new chunk: The chunk is not found");
+
+                            debug_assert!(
+                                matches!(self.chunks.get(next_chunk_index - 1), Some((p, _)) if p == previous),
```

Comment:
`next_chunk_index` should always be >0 too, right? Just making sure this subtract is safe…

Reply:
Yeah correct, if `next` is at position 0, then `previous` must be at position -1 at best, which doesn't exist. Or, alternatively, `Update::NewItemsChunk` or `Update::NewGapChunk` is totally broken, but in this case, we would have other problems 😉.

This subtract is safe.